### ISSUE TITLE
False positives on SAM implicit ServerlessRestApi since 1.54.0: StateMachine Api events ignored, and .Stage checked as a GetAtt

### ISSUE THREAD

**Issue by netfalo** [aws-cloudformation/cfn-lint#4714]
### CloudFormation Lint Version

1.56.3 (regression introduced in 1.54.0; 1.53.3 is clean)

### What operating system are you using?

Linux

### Describe the bug

Two related false positives in the SAM implicit-resource injection added by #4491 ("Remove SAM translator dependency, validate SAM templates via schemas"). Both are regressions in 1.54.0.

**1. `AWS::Serverless::StateMachine` does not trigger implicit `ServerlessRestApi` creation.**

A `StateMachine` with an `Api` event and no `RestApiId` causes SAM to generate `ServerlessRestApi`, exactly as a `Function` does. cfn-lint does not, so any reference to it reports E6001/E6004/E6101.

In `cfnlint/context/context.py`, `_inject_sam_implicit_resources` gates on three resource types:

```
633:  if resource_type == "AWS::Serverless::Function":
694:  if resource_type == "AWS::Serverless::Api":
709:  if resource_type == "AWS::Serverless::HttpApi":
```

`AWS::Serverless::StateMachine` is absent, so its `Events` are never scanned and `needs_rest_api` is never set.

**2. `${ServerlessRestApi.Stage}` is validated as a GetAtt against `AWS::ApiGateway::RestApi`.**

This one reproduces with a plain `Function` too, where the implicit API *is* injected:

```
E6101 'Stage' is not one of ['RestApiId', 'RootResourceId'] in ['us-east-1']
```

`.Stage` is a SAM [resource reference](https://github.com/aws/serverless-application-model/blob/master/docs/internals/generated_resources.rst#api), resolved by the transform before CloudFormation sees it — the same class of construct as `MyFunction.Version` / `MyFunction.Alias`, which cfn-lint does register. The injector adds a sibling resource named `ServerlessRestApiStage` rather than making `Stage` a referable property of `ServerlessRestApi` (and the real translator names that resource `ServerlessRestApiProdStage`).

Confirmed against `aws-sam-translator` on both templates below — it generates `ServerlessRestApi`, `ServerlessRestApiDeployment*` and `ServerlessRestApiProdStage` in each case, and rewrites the output to `${ServerlessRestApiProdStage}`:

```
== t-func.yaml (Function)
   ServerlessRestApi present: True
   Outputs: {"U": {"Value": {"Fn::Sub": "https://${ServerlessRestApi}.execute-api.${AWS::Region}.amazonaws.com/${ServerlessRestApiProdStage}/"}}}
== t-sm.yaml (StateMachine)
   ServerlessRestApi present: True
   Outputs: {"U": {"Value": {"Fn::Sub": "https://${ServerlessRestApi}.execute-api.${AWS::Region}.amazonaws.com/${ServerlessRestApiProdStage}/"}}}
```

`sam validate` also reports both templates valid.

### Expected behavior

Both templates lint clean, as they do on 1.53.3.

Bisect (the two templates below):

| version | Function | StateMachine |
|---|---|---|
| 1.53.3 | clean | clean |
| 1.54.0 | E6101 | E6001, E6004, E6101 |
| 1.56.3 | E6101 | E6001, E6004, E6101 |

### Reproduction template

The two differ only in the resource type.

**Function — reproduces bug 2 only:**

```yaml
AWSTemplateFormatVersion: "2010-09-09"
Transform: AWS::Serverless-2016-10-31
Resources:
  F:
    Type: AWS::Serverless::Function
    Properties:
      Runtime: python3.12
      Handler: app.handler
      InlineCode: "def handler(e,c): pass"
      Events:
        ApiEvent:
          Type: Api
          Properties:
            Path: /
            Method: POST
Outputs:
  U:
    Value: !Sub "https://${ServerlessRestApi}.execute-api.${AWS::Region}.amazonaws.com/${ServerlessRestApi.Stage}/"
```

**StateMachine — reproduces bugs 1 and 2:**

```yaml
AWSTemplateFormatVersion: "2010-09-09"
Transform: AWS::Serverless-2016-10-31
Resources:
  SM:
    Type: AWS::Serverless::StateMachine
    Properties:
      Definition:
        StartAt: X
        States:
          X:
            Type: Pass
            End: true
      Events:
        ApiEvent:
          Type: Api
          Properties:
            Path: /
            Method: POST
Outputs:
  U:
    Value: !Sub "https://${ServerlessRestApi}.execute-api.${AWS::Region}.amazonaws.com/${ServerlessRestApi.Stage}/"
```

Dropping `.Stage` from the Function template (hardcoding the stage) makes it lint clean on 1.56.3, which isolates bug 2 from bug 1.
